### PR TITLE
Fix data races in RefreshClient

### PR DESCRIPTION
**Pull request by lvfxx**
Implementation of #237 
Fixed data race between refresh goroutines and close calls/new requests from heal. In some situations there wasn't enough to call `timer.Stop()` to really stop refresh goroutine. I've added context for each refresh and moved timer creation to synced block to fix it. 
Also fixed invalid second request initialization in `TestNewClient_StopRefreshAtAnotherRequest`, that also led to data race. 
Harmless refactoring.

### CODE REVIEW

File: pkg/networkservice/common/refresh/client.go [hunk 1]
```diff
@@ -35,16 +35,18 @@ import (
...
 
 // NewClient - creates new NetworkServiceClient chain element for refreshing connections before they timeout at the
 // endpoint
 func NewClient() networkservice.NetworkServiceClient {
 	rv := &refreshClient{
-		connectionTimers: make(map[string]*time.Timer),
-		executor:         serialize.Executor{},
+		connectionTimers:       make(map[string]*time.Timer),
+		timerContextCancellers: make(map[string]func()),
+		executor:               serialize.Executor{},
```
Comment (denis-tingajkin) [inline pkg/networkservice/common/refresh/client.go:49]: We do not need to assign value for field `executor`.

File: pkg/networkservice/common/refresh/client.go [hunk 2]
```diff
@@ -59,15 +61,24 @@ func (t *refreshClient) Request(ctx context.Context, request *networkservice.Net
...
 	req.Connection = rv
 	// Setup the timer with the req containing the returned connection
-	ct, err := t.createTimer(ctx, req, opts...)
+
+	expire, err := t.getExpireDuration(request)
 	if err != nil {
 		return nil, errors.Wrapf(err, "Error creating timer from Request.Connection.Path.PathSegment[%d].ExpireTime", request.GetConnection().GetPath().GetIndex())
 	}
 	t.executor.AsyncExec(func() {
+		if ctx.Err() != nil {
```
Comment (denis-tingajkin) [inline pkg/networkservice/common/refresh/client.go:70]: Why do we need to check request context here?

Comment (lvfxx) [reply pkg/networkservice/common/refresh/client.go:70]: For each scheduled refresh request (created in `createTimer`) we set cancel function and save it to the map. When a new request from heal or stop call with the same `connId` will reach `AsyncExec` block, it cancels refresh's context. Then when refresh will try to schedule its next call, it sees that context was cancelled and terminates itself. This check must be placed in `AsyncExec` block to ensure that new request and refresh goroutines will not get access to context concurrently.

File: pkg/networkservice/common/refresh/client.go (hunk 1 above)
Comment (lvfxx) [reply pkg/networkservice/common/refresh/client.go:49]: I'll remove it, thanks.

File: pkg/networkservice/common/refresh/client.go [hunk 3]
```diff
@@ -78,23 +89,31 @@ func (t *refreshClient) Close(ctx context.Context, conn *networkservice.Connecti
 			timer.Stop()
 			delete(t.connectionTimers, conn.GetId())
 		}
+		if canceller, ok := t.timerContextCancellers[conn.GetId()]; ok {
+			canceller()
+			delete(t.timerContextCancellers, conn.GetId())
```
Comment (denis-tingajkin) [inline pkg/networkservice/common/refresh/client.go:94]: Should we also delete the entry from the map connectionTimers?

File: pkg/networkservice/common/refresh/client.go (hunk 2 above)
Comment (denis-tingajkin) [reply pkg/networkservice/common/refresh/client.go:70]: As I can see here using the request context. Should to use a special `refresh context` for refresh goals?
For example you need to create a refreshing context on first Request call and reuse this context in the refresh timer. Is it make sense?

File: pkg/networkservice/common/refresh/client.go (hunk 3 above)
Comment (lvfxx) [reply pkg/networkservice/common/refresh/client.go]: It is deleted two lines above

File: pkg/networkservice/common/refresh/client.go [hunk 4]
```diff
@@ -59,25 +59,37 @@ func (t *refreshClient) Request(ctx context.Context, request *networkservice.Net
...
-			canceller()
+		// check if it is refresh request
+		if refreshCtx := refreshContext(ctx); refreshCtx != nil {
+			// refresh was canceled
+			if refreshCtx.Err() != nil {
+				return
+			}
+			// we reuse non-canceled refresh context for the next refresh request
+			timer := t.createTimer(ctx, req, expire, opts...)
+			t.connectionTimers[req.GetConnection().GetId()] = timer
```
Comment (denis-tingajkin) [inline pkg/networkservice/common/refresh/client.go:76]: [minor] `req.GetConnection().GetId()` reused many times, you can create local variable for it. 
```suggestion
			id := req.GetConnection().GetId()
			t.connectionTimers[id] = timer
```

**Pull request merged**